### PR TITLE
(GRPO) Fix SAPO on version TRL > 0.27.2 and add Luspo from TRL >= 0.28.0

### PR DESCRIPTION
**Pull request by marcandrelarochelle**
- At TRL 0.27.2, the code for the SAPO loss changes and breaks it for 0.27.2 >= for Unsloth users, this aligns the code with TRL and fixes it.
- Luspo was added in TRL 0.28.0, this adds it to Unsloth as well.

### CODE REVIEW

**Comment by danielhanchen** [comment]
Thanks for this. I checked both halves against the real TRL sources (0.27.2, 0.28.0, 1.7.1). The Luspo half is correct; the SAPO rewrite has a correctness bug.

## Luspo: correct

Adding `luspo` to the clipped-surrogate branch matches TRL (`grpo_trainer.py` in 0.28.0 and 1.7.1 groups it with `["grpo", "bnpo", "dr_grpo", "dapo", "luspo"]`), and the new reduction

```python
elif loss_type == "luspo":
    loss = (loss_i * mask.sum(1, keepdim=True)).mean()
    loss = loss / current_gradient_accumulation_steps
```

matches TRL's luspo reduction verbatim (train mode). Good.

## SAPO: missing `-... * advantages` (please fix before merge)

The new formula for `soft_coef_1` is right, but the branch assigns it straight to `loss_i` and drops the final step. TRL 1.7.1 (`grpo_trainer.py:2754-2757`), same in 0.28.0:

```python
temperatures = torch.where(advantages > 0, self.args.sapo_temperature_pos, self.args.sapo_temperature_neg)
soft_coef_1 = torch.sigmoid(temperatures * (coef_1 - 1)) * 4 / temperatures
per_token_loss = -soft_coef_1 * advantages
```

The PR keeps the first two lines but removes the old trailing `loss_i = -loss_i * advantages` without replacing it, so `loss_i = soft_coef_1`. That loses the advantage weighting (the advantage is used only via `advantages > 0` to pick the temperature, then discarded) and flips the sign. Nothing downstream reapplies it (only `off_policy_mask`, the vLLM IS ratio, and `+ beta*kl_i` come after). Net effect: SAPO optimizes the wrong objective.

Fix:

```python
elif loss_type == "sapo":
    temperatures = torch.where(advantages > 0, sapo_temperature_pos, sapo_temperature_neg)
    soft_coef_1 = torch.sigmoid(temperatures * (coef_1 - 1)) * 4 / temperatures
    loss_i = -soft_coef_1 * advantages
```

I confirmed the same `-... * advantages` in TRL 0.27.2 (`:2341`, `per_token_loss = -per_token_loss * advantages`), so this holds across the versions you are targeting. Dropping `get_sapo_token_loss` for the sapo path is fine (TRL 1.7.1 no longer defines it), it is just harmless leftover in the kwargs plumbing now.

**Comment by danielhanchen** [comment]
Pushed the SAPO fix directly (de51d40): restore `loss_i = -soft_coef_1 * advantages` so SAPO keeps the advantage sign and magnitude, matching TRL. Luspo left as-is since it already matches. Thanks again.

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/rl_replacements.py [hunk 1]
```diff
@@ -544,6 +533,10 @@ def grpo_compute_loss(
     elif loss_type in ["cispo", "dapo", "vespo"]:
         normalizer = num_items_in_batch/ num_processes
         loss = (loss_i * mask).sum() / normalizer
+    elif loss_type == "luspo":
+        loss = (loss_i * mask.sum(1, keepdim=True)).mean()
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/rl_replacements.py:537]: This mirrors TRL 1.7.1's luspo reduction verbatim: TRL uses the same (per_token_loss * mask.sum(1, keepdim=True)).mean() and only warns (does not enforce) when importance_sampling_level is not sequence. Keeping parity with TRL here; the sequence-level requirement for luspo is governed by TRL's config, and diverging would break that parity.

**Comment by danielhanchen** [comment]
@codex review

**Review by danielhanchen — APPROVED** [review]
Independently re-reviewed against TRL 0.26.0 through 1.7.1: the SAPO soft_coef_1 formula and Luspo normalizer are identical to TRL's own implementation, the refactor away from the removed get_sapo_token_loss is a verified numerical no-op on older TRL, and gradients flow correctly. Clean.
